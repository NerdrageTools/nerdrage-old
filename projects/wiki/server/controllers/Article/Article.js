export default class Article {
  static defaults = {
    aliases: [],
    campaign: '',
    data: {},
    slug: '',
    tags: [],
    title: '',
  }

  static findOne = async (db, { campaign, slug }) => {
    const data = db.collection('articles').findOne({ campaign, slug })
    return new Article(db, data)
  }

  db = null

  constructor(db, data) {
    this.db = db
    this.data = { ...Article.defaults, ...data }
  }
}
