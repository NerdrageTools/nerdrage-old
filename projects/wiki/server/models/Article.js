import mongoose from 'mongoose'

const Slug = {
  match: [/^[a-z0-9_-]{1,}$/, 'Invalid Slug: only lowercase letters/numbers, _ and -'],
  trim: true,
  type: String,
}

export default mongoose.model('Article', new mongoose.Schema({
  _id: { ...Slug, alias: 'slug' },
  aliases: [{ ...Slug, ref: 'Article' }],
  campaign: { ref: 'Campaign', type: mongoose.Schema.Types.ObjectId },
  createdBy: { ref: 'User', type: mongoose.Schema.ObjectId },
  data: Object,
  editors: [{ ref: 'User', type: mongoose.Schema.ObjectId }],
  lastUpdatedBy: { ref: 'User', type: mongoose.Schema.ObjectId },
  tags: [{ ...Slug, ref: 'Article' }],
  title: { required: true, type: String },
}, {
  timestamps: true,
  versionKey: 'version',
}))
