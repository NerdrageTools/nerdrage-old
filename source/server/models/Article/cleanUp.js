import cheerio from 'cheerio'
import parseHref from '@/utilities/parseHref'

export default function cleanUp(html) {
  const $ = cheerio.load(html || '', { decodeEntities: false, xmlMode: true })
  $('a').each((_, el) => {
    const $el = $(el)
    const href = $el.attr('href')

    if (!href.includes('://')) {
      const parsed = parseHref(href)
      const slug = $el.attr('slug') || parsed.slug
      const type = $el.attr('type') || parsed.type || 'article'
      const subdomain = $el.attr('subdomain') || parsed.subdomain
      $el.attr('href', `${subdomain}/${type}/${slug}`)
    }

    $el
      .removeClass('cross-campaign')
      .removeClass('external')
      .removeClass('missing')
      .removeClass($el.attr('type'))
      .removeAttr('slug')
      .removeAttr('subdomain')
      .removeAttr('target')
      .removeAttr('type')

    if (($el.attr('class') || '').trim() === '') {
      $el.removeAttr('class')
    }
  })
  $('include').each((_, el) => $(el).removeClass('noedit').html(''))

  return $.html().trim()
}
