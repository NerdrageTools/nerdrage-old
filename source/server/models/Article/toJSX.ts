import { decodeHTML, decodeXML, encodeHTML5 } from 'entities'
import { html_beautify as beautify } from 'js-beautify'
import { VOID_ELEMENTS } from '~/utilities/VOID_ELEMENTS'

const BEAUTIFY_OPTIONS = {
	end_with_newline: true,
	indent_char: ' ',
	indent_size: 2,
	preserve_newlines: false,
	wrap_line_length: 0,
}
const VOID_ELEMENT_CLOSING_TAGS = new RegExp(`<\/(${VOID_ELEMENTS.join('|')})>`, 'g')

export const toJSX = (html: string = '') => (
	([
		decodeHTML,
		decodeXML,
		(value: string) => value.replace(VOID_ELEMENT_CLOSING_TAGS, '')
	]).reduce((value, fn) => fn(value), html)
)
