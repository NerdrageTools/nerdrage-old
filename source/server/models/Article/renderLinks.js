import url from 'url'
import cheerio from 'cheerio'
import Article from '@/server/models/Article'
import unique from '@/utilities/unique'

export default async function renderLinks(html, campaignFilter = { campaign: null }) {
	const $ = cheerio.load(html || '', { decodeEntities: false, xmlMode: true })
	const links = []

	$('a').each((_, element) => {
		const $link = $(element)
		const href = $link.attr('href') || ''
		const parsedUrl = url.parse(href)

		const isInternal = !parsedUrl.hostname
		if (isInternal) {
			const parts = parsedUrl.pathname.split('/').filter(Boolean)
			const slug = parts.pop()
			const type = parts.pop() || 'article'
			const subdomain = parts.pop() || ''

			$link.attr('href', `/${type}/${slug}`)
			$link.attr('slug', slug)
			$link.attr('subdomain', subdomain)
			$link.attr('type', type)

			if (subdomain) {
				$link.addClass('cross-campaign')
			} else {
				links.push(slug)
			}
		} else {
			$link.attr('target', '_new').addClass('external')
		}
	})

	const linked = await Article.find({
		$and: [
			campaignFilter,
			{ $or: [{ slug: { $in: links } }, { aliases: { $in: links } }] },
		],
	}).select('slug aliases')
		.exec()
	const foundLinks = (linked || []).reduce((all, { aliases, slug }) => (
		[...all, slug, ...aliases]
	), [])

	const missing = links.filter(link => !foundLinks.includes(link))

	$('a').each((_, element) => {
		const $link = $(element)
		const slug = $link.attr('slug') || ''
		const type = $link.attr('type') || ''
		if (type === 'article' && missing.includes(slug)) {
			$link.addClass('missing')
		}
	})

	return Promise.resolve({
		html: $.html(),
		links: unique(links).sort(),
		missing: unique(missing).sort(),
	})
}
