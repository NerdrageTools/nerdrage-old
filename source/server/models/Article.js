import cheerio from 'cheerio'
import entities from 'entities'
import { html_beautify as beautify } from 'js-beautify'
import mongoose from 'mongoose'
import Slug from '@/server/models/Slug'
import computeSearchKeys from '@/utilities/computeSearchKeys'
import createCampaignFilter from '@/utilities/createCampaignFilter'
import transclude from '@/utilities/transclude'
import unique from '@/utilities/unique'
import renderLinks from './Article/renderLinks'
import cleanUp from './Article/cleanUp'

const { ObjectId: ObjectIdType } = mongoose.Schema.Types
const slugifyArray = array => unique(array.map(value => (
	value.toLowerCase().replace(/[^\w-_]/g, '-').replace(/-{2,}|^-|-$/g, '')
))).sort()
const BEAUTIFY_OPTIONS = {
	end_with_newline: true,
	indent_char: ' ',
	indent_size: 2,
	preserve_newlines: false,
	wrap_line_length: 0,
}

export const ArticleSchema = new mongoose.Schema({
	aliases: [Slug],
	campaign: { default: null, ref: 'Campaign', required: false, type: ObjectIdType, unique: false },
	createdBy: { ref: 'User', type: ObjectIdType },
	data: Object,
	html: { default: '', type: String },
	lastUpdatedBy: { ref: 'User', type: ObjectIdType },
	plainText: { default: '', select: false, type: String },
	searchKeys: { select: false, type: [String] },
	secret: { default: false, type: Boolean },
	slug: { ...Slug, required: true, unique: false },
	tags: [Slug],
	template: { default: false, type: Boolean },
	title: { required: true, type: String },
}, {
	id: false,
	timestamps: true,
	toJSON: {
		transform(_, returnValue) {
			/* eslint-disable no-param-reassign */
			delete returnValue.plainText
			delete returnValue.searchKeys
			/* eslint-enable no-param-reassign */
		},
		virtuals: true,
	},
	toObject: { virtuals: true },
	versionKey: 'version',
})

ArticleSchema.index({ campaign: 1, slug: 1 }, { unique: true })
ArticleSchema.index({ plainText: 'text', title: 'text' })
ArticleSchema.index({ searchKeys: 1 })
ArticleSchema.pre('save', function (next) {
	this.aliases = slugifyArray(this.aliases)
	this.html = beautify(cleanUp(this.html), BEAUTIFY_OPTIONS)
	this.tags = slugifyArray(this.tags)

	const $ = cheerio.load(this.html, { decodeEntities: false, xmlMode: true })
	$('br').replaceWith('\r\n')
	this.plainText = entities.decodeHTML($.text().replace(/[\r\n]/g, ' ').replace(/\s+/g, ' '))
	this.searchKeys = computeSearchKeys(this.plainText)

	mongoose.models.Article.updateMany(
		{ _id: { $ne: this._id }, campaign: this.campaign },
		{ $pull: { aliases: { $in: this.aliases } } },
	).then(() => next())
})

ArticleSchema.methods.render = async function (campaign) {
	const campaignFilter = createCampaignFilter(campaign)
	const includes = await transclude(this.html, campaignFilter)
	const links = await renderLinks(includes.html, campaignFilter)
	const childArticles = await mongoose.models.Article
		.aggregate([
			{ $match: {
				$and: [campaignFilter, { tags: { $in: [this.slug, ...this.aliases] } }],
				template: false,
			} },
			{ $group: {
				_id: '$slug',
				id: { $first: '$_id' },
				slug: { $first: '$slug' },
				title: { $first: '$title' },
			} },
			{ $project: { _id: '$id', slug: '$slug', title: '$title' } },
		])
		.exec()

	const html = beautify(links.html, BEAUTIFY_OPTIONS)

	return Promise.resolve({
		...this.toJSON(),
		childArticles,
		html,
	})
}

const Article = mongoose.model('Article', ArticleSchema)

export default Article
