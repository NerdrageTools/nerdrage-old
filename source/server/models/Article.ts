import {
	getModelForClass, index, modelOptions, pre, prop, Ref, ReturnModelType, Severity,
} from '@typegoose/typegoose'
import cheerio from 'cheerio'
import * as entities from 'entities'
import { html_beautify as beautify } from 'js-beautify'
import { Slug } from '~/server/models/Slug'
import { computeSearchKeys } from '~/utilities/computeSearchKeys'
import { createCampaignFilter } from '~/utilities/createCampaignFilter'
import { transclude } from '~/utilities/transclude'
import { unique } from '~/utilities/unique'
import { cleanUp } from '~/server/models/Article/cleanUp'
import { renderLinks } from '~/server/models/Article/renderLinks'
import { Campaign } from './Campaign'
import { User } from './User'

const slugifyArray = (array: string[]) => unique(array.map(value => (
	value.toLowerCase().replace(/[^\w-_]/g, '-').replace(/-{2,}|^-|-$/g, '')
))).sort()
const BEAUTIFY_OPTIONS = {
	end_with_newline: true,
	indent_char: ' ',
	indent_size: 2,
	preserve_newlines: false,
	wrap_line_length: 0,
}

export interface IArticleData {
	_id?: string,
	aliases?: string[],
	campaign: Ref<typeof Campaign>,
	createdAt?: Date,
	createdBy?: Ref<User>,
	html?: string,
	lastUpdatedBy?: Ref<User>,
	plainText?: string,
	secret?: boolean,
	slug: string,
	tags?: string[],
	template?: boolean,
	title: string,
	updatedAt?: Date,
	version?: number,
}

export interface IArticleLink {
	slug: string,
	subdomain: string,
	title: string,
}

export interface IArticleSearchResult {
	nameHits: number,
	preview: string[],
	secret: boolean,
	slug: string,
	subdomain: string,
	textHits: number,
	title: string,
}

@index({ campaign: 1, slug: 1 }, { unique: true })
@index({ plainText: 'text', title: 'text' })
@modelOptions({
	options: { allowMixed: Severity.ALLOW },
	schemaOptions: { id: false, timestamps: true, versionKey: 'version' },
})
@pre<Article>('save', async function (next) {
	this.aliases = slugifyArray(this.aliases)
	this.html = beautify(cleanUp(this.html), BEAUTIFY_OPTIONS)
	this.tags = slugifyArray(this.tags)

	const $ = cheerio.load(this.html, { decodeEntities: false, xmlMode: true })
	$('br').replaceWith('\r\n')
	// @ts-expect-error - .text() is perfectly valid on Root
	this.plainText = entities.decodeHTML($.text().replace(/[\r\n]/g, ' ').replace(/\s+/g, ' '))
	this.searchKeys = computeSearchKeys(this.plainText)

	await Articles.updateMany(
		{ _id: { $ne: this._id }, campaign: this.campaign },
		{ $pull: { aliases: { $in: this.aliases } } },
	)
	next()
})
export class Article implements IArticleData {
	_id?: string
	createdAt?: Date
	updatedAt?: Date
	version?: number

	@prop({ default: [], type: [String] })
	public aliases: string[] = []

	@prop({ default: null, ref: 'Campaign' })
	public campaign: Ref<typeof Campaign>

	@prop({ ref: 'User' })
	public createdBy?: Ref<User>

	@prop({ default: '' })
	public html: string = ''

	@prop({ ref: 'User' })
	public lastUpdatedBy?: Ref<User>

	@prop({ default: '' })
	public plainText: string = ''

	@prop({ default: false })
	public secret: boolean = false

	@prop({ ...Slug, required: true })
	public slug: string = ''

	@prop({ default: [], type: [String] })
	public tags: string[] = []

	@prop({ default: false })
	public template: boolean = false

	@prop({ required: true })
	public title: string = ''

	async render(campaign?: ICampaign): Promise<Article> {
		const campaignFilter = createCampaignFilter(campaign)
		const includes = await transclude(this.html, campaignFilter)
		const links = await renderLinks(includes.html, campaignFilter)
		const childArticles = await Articles.aggregate([
			{ $match: {
				$and: [campaignFilter, { tags: { $in: [this.slug, ...this.aliases] } }],
				template: false,
			} },
			{ $group: {
				_id: '$slug',
				id: { $first: '$_id' },
				slug: { $first: '$slug' },
				title: { $first: '$title' },
			} },
			{ $project: { _id: '$id', slug: '$slug', title: '$title' } },
		]).exec()

		const html = beautify(links.html, BEAUTIFY_OPTIONS)
		return Promise.resolve({ ...this.toJSON(), childArticles, html })
	}
}

export const Articles = getModelForClass(Article)
