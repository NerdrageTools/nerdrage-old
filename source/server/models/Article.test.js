import mongoose from 'mongoose'
import Article from './Article'

const { ObjectId } = mongoose.Types

describe('models/Article', () => {
  beforeAll(async done => {
    await mongoose.connect('mongodb://localhost/test-articles', {
      useCreateIndex: true,
      useNewUrlParser: true,
    })
    done()
  })
  beforeEach(async done => {
    await Article.deleteMany({})
    await Article.create([
      { _id: ObjectId('aaaaaaaaaaaa'), aliases: ['albatross'], slug: 'a', title: 'A' },
      { _id: ObjectId('bbbbbbbbbbbb'), aliases: ['boggin', 'bug'], slug: 'b', title: 'B' },
      { _id: ObjectId('cccccccccccc'), aliases: ['carpet', 'cactus'], slug: 'c', title: 'C' },
    ])
    done()
  })

  describe('onSave', () => {
    it('uniques and sorts aliases/tags', async done => {
      await Article.deleteMany({})

      const article = new Article({
        _id: ObjectId('5c6f906d4a0e252061614c8c'),
        aliases: ['b', 'b', 'c', 'a'],
        slug: 'slug',
        tags: ['b', 'b', 'c', 'a'],
        title: 'foo',
      })
      let updated = await article.save()

      expect(updated.toJSON().aliases).toEqual(['a', 'b', 'c'])
      expect(updated.toJSON().tags).toEqual(['a', 'b', 'c'])

      updated.aliases.push('a')
      updated.tags.push('a')

      expect(updated.toJSON().aliases).toEqual(['a', 'b', 'c', 'a'])
      expect(updated.toJSON().tags).toEqual(['a', 'b', 'c', 'a'])

      updated = await updated.save()

      expect(updated.toJSON().aliases).toEqual(['a', 'b', 'c'])
      expect(updated.toJSON().tags).toEqual(['a', 'b', 'c'])

      await updated.remove({ _id: ObjectId('5c6f906d4a0e252061614c8c') })
      done()
    })

    it('clears aliases from other articles', async done => {
      const NEW = await Article.create({
        aliases: ['albatross', 'bug', 'carpet'],
        slug: 'new',
        title: 'New',
      })

      const A = await Article.findOne({ slug: 'a' })
      const B = await Article.findOne({ slug: 'b' })
      const C = await Article.findOne({ slug: 'c' })
      expect(A.toJSON().aliases).not.toContain('albatross')
      expect(B.toJSON().aliases).not.toContain('bug')
      expect(C.toJSON().aliases).not.toContain('carpet')
      expect(NEW.toJSON().aliases).toEqual(['albatross', 'bug', 'carpet'])

      done()
    })
  })

  it('enforces unique aliases onSave', async () => {
    expect(await Article.find({ aliases: 'albatross' })).toHaveLength(1)
    expect(await Article.find({ aliases: 'bug' })).toHaveLength(1)
    expect(await Article.find({ aliases: 'carpet' })).toHaveLength(1)

    const article = await new Article({
      _id: ObjectId('dddddddddddd'),
      aliases: ['albatross', 'bug', 'carpet'],
      html: `
        <a href="/a">A</a>
        <a href="/b">B</a>
        <a href="/e">E</a>
        <a href="/f">F</a>
      `,
      slug: 'd',
      title: 'D',
    }).save()

    expect(article.toJSON().aliases).toEqual(['albatross', 'bug', 'carpet'])
    const albatross = await Article.find({ aliases: 'albatross' })
    expect(albatross).toHaveLength(1)
    expect(albatross[0]._id.equals(article._id)).toBeTruthy()

    const bug = await Article.find({ aliases: 'bug' })
    expect(bug).toHaveLength(1)
    expect(bug[0]._id.equals(article._id)).toBeTruthy()

    const carpet = await Article.find({ aliases: 'carpet' })
    expect(carpet).toHaveLength(1)
    expect(carpet[0]._id.equals(article._id)).toBeTruthy()

    return Article.deleteMany({ slug: { $in: ['a', 'b', 'c'] } })
  })

  it('renders cleanly', async done => {
    const NEW = new Article({
      html: `
        <h1>New!</h1>
        <a href="/a">A</a>
        <a href="/missing-link">Missing!</a>
        <include from="b" sections="*" />
        <include from="missing-include" sections="*" />
      `,
      slug: 'new',
      title: 'New',
    })

    const rendered = await NEW.render()
    expect(rendered.links).toEqual(['a', 'b', 'missing-include', 'missing-link'])
    expect(rendered.missing).toEqual(['missing-include', 'missing-link'])
    expect(rendered.html).toMatchSnapshot()

    done()
  })

  afterAll(async done => {
    await mongoose.connection.db.dropDatabase()
    await mongoose.disconnect()
    done()
  })
})
