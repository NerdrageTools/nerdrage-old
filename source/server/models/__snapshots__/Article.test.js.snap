// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/Article renders cleanly 1`] = `
"
        <h1>New!</h1>
        <a href=\\"/page/a\\">A</a>
        <a href=\\"/page/missing-link\\" class=\\"missing\\">Missing!</a>
        <include from=\\"b\\" sections=\\"*\\" class=\\"noedit\\">
  
</include>
        <include from=\\"missing-include\\" sections=\\"*\\" class=\\"noedit\\">
<div class=\\"warning\\">Included Article 'missing-include' does not exist</div>
</include>
      "
`;
